wgpu-core/src/weak_vec.rs

```diff
@@ -0,0 +1,67 @@
+//! Module containing the [`WeakVec`] API.
+
+use std::sync::Weak;
+
+/// A container that holds Weak references of T.
+///
+/// On `push` it scans its contents for weak references with no strong references still alive and drops them.
+#[derive(Debug)]
+pub(crate) struct WeakVec<T> {
+    inner: Vec<Option<Weak<T>>>,
+}
+
+impl<T> Default for WeakVec<T> {
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+        }
+    }
+}
+
+impl<T> WeakVec<T> {
+    pub(crate) fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Pushes a new element to this collection, dropping older elements that no longer have
+    /// a strong reference to them.
+    ///
+    /// NOTE: The length and capacity of this collection do not change when old elements are
+    /// dropped.
+    pub(crate) fn push(&mut self, value: Weak<T>) {
```

Comment:
question(non-blocking): Having an API that _only_ exposes a `push` and an `IntoIterator` interface makes it easy to ask: How are we ensuring that a `WeakVec` gets its capacity shrunk down again if it ever gets large? I presume that this is done by consuming the `WeakVec`s with `IntoIterator`.

We had this problem to consider before this PR was filed, so I won't block merging with this question. I intend to read code to answer this, but I want to confirm this understanding "out loud", as it were. We can punt to an issue, if we discover something substantial with this discussion.

Reply:
I would imagine it's not common for apps to create lots of bind groups and texture views then consistently use less over time. `Weak` is also not large at 8 bytes.